mg.c

```diff
@@ -3367,6 +3367,16 @@ Perl_magic_set(pTHX_ SV *sv, MAGIC *mg)
         else sv_setiv(mg->mg_obj, (IV)PerlProc_getpid());
         break;
     case '0':
+        if (!sv_utf8_downgrade(sv, /* fail_ok */ TRUE)) {
```

Comment:
It makes more sense to me to move this code into `S_set_dollarzero`

Reply:
That’s where I had it, but then I realized that I need to update $0 itself. Without that update, `$0 = "\x{100}"` will cause $0 to have code point 256 while the actual process name is the 2 bytes of that code point in UTF-8.